### ISSUE TITLE
grafana support for nanosecond resolution

### ISSUE THREAD

**Issue by bf228** [grafana/grafana#6252]
I've got data in microsecond resolution:

t=[1475652098.160606] msg-time=[2016-10-05 07:21:38.160606] sn=[6111]
t=[1475652359.457696] msg-time=[2016-10-05 07:25:59.457696] sn=[6112]
t=[1475652359.457733] msg-time=[2016-10-05 07:25:59.457733] sn=[6113]
t=[1475652359.457737] msg-time=[2016-10-05 07:25:59.457737] sn=[6114]
t=[1475652359.457741] msg-time=[2016-10-05 07:25:59.457741] sn=[6115]
t=[1475652359.457741] msg-time=[2016-10-05 07:25:59.457741] sn=[6116]

and grafana defaults to using ms resolution.   The end result is that while influxdb returns all of these values to grafana, their timestamps are truncated to millisecond values and so only the first and last samples are processed by grafana.

Is there a way to get grafana to use nanosecond resolution?  (I'll slightly skew my samples's timestamps from microsecond resolution to nanosecond resolution so each sample has a unique nanosecond timestamp.

thanks

**Comment by torkelo** [comment]
No there is no way to do that. It would be quite tricky as all time formats in javascript (and time libs) only go down to millisecond resolution.

**Comment by bf228** [comment]
Thanks.  Since I want grafana to process all points, what if I did the following.  What if I created a new field in the influxdb lineprotocol which was the unique ns value.  The time would still be limited to millisecond resolution but since each sample would have a unique field value, would grafana treat them unique?  So something like (not using real time values obviously):

key,foo="bar",ns=4200 value=x 420
key,foo="bar"ns=4201  value=y 420
key,foo="bar",ns=4202 value=z 420

etc?

Would this result in all data points being processed by grafana?

**Comment by emannion** [comment]
Hi @torkelo. Have you any update on this nanosecond resolution feature?  It is something my organization could use at the moment. Thanks.

**Comment by rtmaster** [comment]
+1 on nanosecond resolution

**Comment by deepeshreja** [comment]
+1

**Comment by PeterHuewe** [comment]
+1

**Comment by comcomservices** [comment]
$1,000.00 if anyone wants to update grafana to support ns timescale resolution.  Our use case requires importing graph data from influxdb and displaying it.

Contact me if interested, work must be professional with pull requests for all changes needed

**Comment by wolfig** [comment]
+1
Urgent use case for scientific and machine monitoring: we want to monitor periodic signals from particle accelerators having 100 ns resolution and periods of typically 10 micros.

**Comment by Kazmirchuk** [comment]
+1 
we need to visualize spacecraft data, and all modern spacecraft come with nanosecond-resolution onboard clocks

**Comment by ummarbhutta** [comment]
+1 
We need this feature to show very high resolution data

**Comment by IsmaeelHaider** [comment]
+1

**Comment by christophhosp** [comment]
+1
We also need this feature for high resolution data from production machines.

**Comment by EricDoerr-SL** [comment]
+1
We also need this feature for nanosecond resolution.

**Comment by JoeyyT** [comment]
+1

**Comment by wallishuang** [comment]
+1

**Comment by mattp0** [comment]
+1

**Comment by fifeclub** [comment]
Hard to believe this is still an issue!
+1
RFC3339Nano is supported my moment

**Comment by rhopkinsjr** [comment]
I would like to know how well Grafana supports millisecond resolution. 

We use PostgreSQL with TimeScaleDB enabled. Currently, our data is in UNIX millisec, formatted as an integer like 1475652098160. We have to enter "timestamp/1000" into the time field of the query builder. But many of our users forget to do it or hate to do it. I want a hands-off solution.

If I asked my engineers to have a column labeled "time" with values formatted like 1475652098.160 or 2016-10-05 07:21:38.160 would Grafana work? Specifically:

1. Would Grafana correctly ingest the data, recognizing and preserving the millisec values?
2. Would a non-technical user at the query builder be able to work with millisec data as easily as with sec data? 
3. Would the "time" column be selected by default?
4. Would time functions such as time ($__interval, NULL) be applied correctly in GROUP BY?
Thanks all!

**Comment by pkrishnanreflexion** [comment]
+1

**Comment by MurrMack** [comment]
+1

**Comment by oliora** [comment]
+1

**Comment by beeksanalytics** [comment]
+1

**Comment by tskarhed** [comment]
We have have systems that return nanosecond timestamps. In terms of seeing the values, most cases milliseconds would do. There are a few edge cases (related to hardware timings) that a micro and nano resolution would be very useful.

**Comment by tskarhed** [comment]
I have created a sample dashboard for this https://gist.github.com/tskarhed/4f8dd61d67bf77b20ba1c6bb71684ce8

**Comment by DWebb0** [comment]
+1 for network switching and an increasing number of IoT use cases

**Comment by scandey** [comment]
+1 for space physics instrumentation, this is unfortunately a deal breaker for our adoption of Grafana

**Comment by leeoniya** [comment]
there is some movement on this:

https://github.com/grafana/grafana-plugin-sdk-go/pull/647
https://github.com/grafana/grafana/pull/64529
https://github.com/grafana/grafana/issues/67889

we're also considering adding a `timeOffset` field type. we also have a new Trend panel in v10 that has a numeric x which can be useful here.

even today, in 2023, JavaScript has no native support for higher-than-millisecond precision in date objects, and Number has only enough precision for microsecond epochs, which means dealing with much much slower strings for encoding and processing higher precision.

my thinking is that working with raw nanosecond epochs is generally not very useful. if you were to zoom to 1s of data you really only might need the calendar time of the first datapoint, and the rest can just be millisecond, microsecond, or nanosecond numeric offsets from that, rather than full epochs for each datapoint, which the timeOffset field type is envisioned to handle.

are there common cases where people are expecting to need sub-millisecond resolution at time ranges > 1s, or where the full epoch of each datapoint is actually relevant rather than an offset from a known start?

**Comment by scandey** [comment]
Those options are really promising! It is great to hear that there is movement towards this possibility.

> are there common cases where people are expecting to need sub-millisecond resolution at time ranges > 1s, or where the full epoch of each datapoint is actually relevant rather than an offset from a known start?

In my particular case for particle data, nanosecond offsets from a regularly occurring start point would be perfect. While I could work with x.000 seconds from epoch being the offset start point, my data ingestion process would be much cleaner if I could set the start point to an arbitrary number of seconds and milliseconds from epoch (ie x.838s from epoch, y.676 from epoch, etc.). The actual data points are already nanosecond offsets from these time values, so it would be cleanest just to keep using the exact time values as reportedly

**Comment by ummarbhutta** [comment]
> 

Offset approach would work for us, for now we create a custom webUI where we are displaying data already in terms of time offset on x-axis. Looking forward for similar functionality in Grafana.

**Comment by trungk99** [comment]
> Thanks. Since I want grafana to process all points, what if I did the following. What if I created a new field in the influxdb lineprotocol which was the unique ns value. The time would still be limited to millisecond resolution but since each sample would have a unique field value, would grafana treat them unique? So something like (not using real time values obviously):
> 
> key,foo="bar",ns=4200 value=x 420 key,foo="bar"ns=4201 value=y 420 key,foo="bar",ns=4202 value=z 420
> 
> etc?
> 
> Would this result in all data points being processed by grafana?

Hi @ghost , 
Can you choose me to see how to realize this idea?

**Comment by quanghungb** [comment]
+1

**Comment by Priyanshi3005** [comment]
+1

**Comment by msawers-surest** [comment]
+1

**Comment by oktavianabd** [comment]
+1

**Comment by SebastiaanDenHertog** [comment]
+1

**Comment by geerligs** [comment]
I need microseconds resolution, so +1 from me!

**Comment by rubenvizros** [comment]
+1

**Comment by rwader-swi** [comment]
One can convert nanosecond to milliseconds if your query is not having group by clause. If it is having one then you're the chosen one for troubles

**Comment by SteveRBeeks** [comment]
> One can convert nanosecond to milliseconds if your query is not having group by clause. If it is having one then you're the chosen one for troubles

I'm not sure what you mean, @rwader-swi , I don't think you understand the problem fully - please elaborate.

**Comment by sdistasi** [comment]
+1 for microsecond resolution

**Comment by gtk-grafana** [comment]
For datasources that support `nanos` (mostly log data sources), the `sort by` transformation now supports sorting by nanoseconds in https://github.com/grafana/grafana/pull/108564.

Support for the table viz is being added in https://github.com/grafana/grafana/pull/108614.

Time series panels are not yet supported AFAIK.

**Comment by slotrans** [comment]
+1 for microseconds. Charting a lot of high-precision robotics telemetry and truncating to milliseconds produces incorrect results sometimes

specifically Time Series and State Timeline need higher precision in their time axis processing

surely in 2025 there's a JS time library that supports higher precision?

**Comment by gtk-grafana** [comment]
@slotrans there is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal/Duration/nanoseconds, but we're still waiting on browser support as only Firefox supports it. 

CC: @leeoniya

**Comment by slotrans** [comment]
@gtk-grafana https://js-joda.github.io/js-joda/ appears to support nanoseconds

I'm no JS guy so maybe there's a reason not to use that. It just doesn't seem like _browser support_ should be a blocker here. This can all be done in userspace, even with JS' limitation of 53-bit integers.

**Comment by gtk-grafana** [comment]
While Grafana provides Moment (which is basically sugar on top of js Date) in an attempt to make the lives of app/plugin developers easier, migrating to a new library for nanos support would basically require a rewrite of core Grafana, not to mention every plugin and datasource. Also the dataviz layer has zero dependencies and relies on native js/browser functionality for performance reasons. Since js-joda has its own limitations and drawbacks (like not being interoperable with native browser date stuff), I would assume we don't want to sink potentially thousands of hours into something that would probably need to get replaced with the native browser solution anyway.

Full disclosure I'm not on the teams responsible for this kind of decision, and I'm 100% in support for nanos support in Grafana, but Grafana is an open source ecosystem with date math at the core, so even if it was decided today to switch over to another date library that supported nanos we would probably have native browser support for nanos before we finished replacing moment. 

TL;DR it's way more complicated then you'd think

**Comment by abbbe** [comment]
Here we go, a scatter plot with ns-level resolution.

<img width="1833" height="850" alt="Image" src="https://github.com/user-attachments/assets/bb37a55d-7156-4998-bd32-93260906f674" />

[edge-nanoseconds-dashboard.json](https://github.com/user-attachments/files/24511482/edge-nanoseconds-dashboard.json)

PS. There seems to be no limit to what Claude Code can do, if you push it hard enough ;)
